### PR TITLE
[pkg/ottl] Added 'Adding new functions/converters' guidelines

### PR DESCRIPTION
**Pull request by michalpristas**
**Description:** 
As discussed [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/32433#issuecomment-2117931670) we should agree on some guidelines when it comes to creating proposals for new functions and converters 

cc @TylerHelmuth 

**Link to tracking Issue:** -

**Testing:** -

**Documentation:** -

### CODE REVIEW

File: pkg/ottl/ottlfuncs/README.md [hunk 1]
```diff
@@ -1308,3 +1308,25 @@ Functions should be named and formatted according to the following standards.
 - Functions that interact with multiple items MUST have plurality in the name. Ex: `truncate_all`, `keep_keys`, `replace_all_matches`.
 - Functions that interact with a single item MUST NOT have plurality in the name. If a function would interact with multiple items due to a condition, like `where`, it is still considered singular. Ex: `set`, `delete`, `replace_match`.
 - Functions that change a specific target MUST set the target as the first parameter.
+
+## Adding New Functions/Converters
+
+Before raising a PR with a new function or converter, raise a PR to verify its acceptance. While acceptance is strongly specific to a specific use case, consider these guidelines for early assessment.
```
Comment (evan-bradley) [inline pkg/ottl/ottlfuncs/README.md:1314]:
```suggestion
## Adding New Editors/Converters

Before raising a PR with a new Editor or Converter, raise an issue to describe the proposed function. While acceptance strongly depends on factors specific to each proposal, consider these guidelines for early assessment.
```

File: pkg/ottl/ottlfuncs/README.md [hunk 2]
```diff
@@ -1308,3 +1308,25 @@ Functions should be named and formatted according to the following standards.
...
+
+It will be up for discussion if your proposal solves an issue that can be achieved in another way but does not improve user experience or performance.
+
+Your proposal likely won't be accepted if:
+- User experience is worse and assumes a highly technical user,
+- The performance of your proposal very negatively affects the processing pipeline.
+
+As with code, `ottl` should aim for readability first. This means:
+- Using short, meaningful, and descriptive names,
+- Ensuring naming consistency across functions and converters,
```
Comment (evan-bradley) [inline pkg/ottl/ottlfuncs/README.md:1329]:
```suggestion
- Ensuring naming consistency across Editors and Converters,
```

File: pkg/ottl/ottlfuncs/README.md [hunk 3]
```diff
@@ -1308,3 +1308,25 @@ Functions should be named and formatted according to the following standards.
...
+- Avoiding deep nesting to achieve desired transformations,
+- Ensuring functions and converters have a single responsibility,
```
Comment (evan-bradley) [inline pkg/ottl/ottlfuncs/README.md:1331]:
```suggestion
- Ensuring Editors and Converters have a single responsibility,
```

File: pkg/ottl/ottlfuncs/README.md [hunk 4]
```diff
@@ -1308,3 +1308,25 @@ Functions should be named and formatted according to the following standards.
 - Functions that interact with multiple items MUST have plurality in the name. Ex: `truncate_all`, `keep_keys`, `replace_all_matches`.
 - Functions that interact with a single item MUST NOT have plurality in the name. If a function would interact with multiple items due to a condition, like `where`, it is still considered singular. Ex: `set`, `delete`, `replace_match`.
 - Functions that change a specific target MUST set the target as the first parameter.
+
+## Adding New Functions/Converters
```
Comment (evan-bradley) [inline pkg/ottl/ottlfuncs/README.md:1312]:
```suggestion
## Adding New Editors/Converters
```

**Review by evan-bradley — COMMENTED** [review]
Thanks for adding this!

File: pkg/ottl/ottlfuncs/README.md [hunk 5]
```diff
@@ -1308,3 +1308,25 @@ Functions should be named and formatted according to the following standards.
...
+
+Before raising a PR with a new function or converter, raise a PR to verify its acceptance. While acceptance is strongly specific to a specific use case, consider these guidelines for early assessment.
+
+You're more than welcome to open an issue if:
```
Comment (TylerHelmuth) [inline pkg/ottl/ottlfuncs/README.md:1316]:
```suggestion
Your proposal likely will be accepted if:
```

File: pkg/ottl/ottlfuncs/README.md [hunk 6]
```diff
@@ -1308,3 +1308,25 @@ Functions should be named and formatted according to the following standards.
...
+Your proposal likely won't be accepted if:
+- User experience is worse and assumes a highly technical user,
+- The performance of your proposal very negatively affects the processing pipeline.
+
+As with code, `ottl` should aim for readability first. This means:
+- Using short, meaningful, and descriptive names,
+- Ensuring naming consistency across functions and converters,
+- Avoiding deep nesting to achieve desired transformations,
+- Ensuring functions and converters have a single responsibility,
+- Following the DRY principle, avoiding multiple functions doing the same thing.
```
Comment (TylerHelmuth) [inline pkg/ottl/ottlfuncs/README.md:1332]: I think we can omit this principle considering we state above how important user experience is and we are gonna encounter lots of scenarios where `replace_pattern` could be a solution, but isn't user-friendly.

**Comment by TylerHelmuth** [comment]
@djaglowski I'm very interested in your opinion here, as new functions are similar to new stanza operators.

File: pkg/ottl/ottlfuncs/README.md [hunk 7]
```diff
@@ -1353,3 +1353,24 @@ Functions should be named and formatted according to the following standards.
 - Functions that interact with multiple items MUST have plurality in the name. Ex: `truncate_all`, `keep_keys`, `replace_all_matches`.
 - Functions that interact with a single item MUST NOT have plurality in the name. If a function would interact with multiple items due to a condition, like `where`, it is still considered singular. Ex: `set`, `delete`, `replace_match`.
 - Functions that change a specific target MUST set the target as the first parameter.
+
+## Adding New Editors/Converters
+
+Before raising a PR with a new function or converter, raise a PR to verify its acceptance. While acceptance is strongly specific to a specific use case, consider these guidelines for early assessment.
```
Comment (djaglowski) [inline pkg/ottl/ottlfuncs/README.md:1359]:
```suggestion
Before raising a PR with a new function or converter, raise an Issue to verify its acceptance. While acceptance is strongly specific to a specific use case, consider these guidelines for early assessment.
```

**Review by djaglowski — COMMENTED** [review]
These guidelines seem reasonable to me. I think they leave room for interpretation, which is good because in some cases we may find other factors to be important.

**Review by TylerHelmuth — APPROVED** [review]

File: pkg/ottl/ottlfuncs/README.md [hunk 8]
```diff
@@ -1400,3 +1400,24 @@ Functions should be named and formatted according to the following standards.
...
+- The proposed solution significantly improves user experience and readability for very common use cases,
+- The proposed solution is more performant in cases where it is possible to achieve the same result with existing options.
+
+It will be up for discussion if your proposal solves an issue that can be achieved in another way but does not improve user experience or performance.
+
+Your proposal likely won't be accepted if:
+- User experience is worse and assumes a highly technical user,
+- The performance of your proposal very negatively affects the processing pipeline.
+
+As with code, `ottl` should aim for readability first. This means:
```
Comment (evan-bradley) [inline pkg/ottl/ottlfuncs/README.md:1419]:
```suggestion
As with code, OTTL aims for readability first. This means:
```

**Review by evan-bradley — APPROVED** [review]
Thanks @michalpristas.

**Pull request merged**